lib/tpm2_hmac_auth.c

```diff
@@ -0,0 +1,208 @@
+    bool res;
+    uint8_t temp[4];
+    //Get Entity names from handles
+    switch((entity_handle & TPM2_HR_RANGE_MASK)>>TPM2_HR_SHIFT) {
+        case TPM2_HT_NV_INDEX:
+            res = tpm2_util_nv_read_public(sapi_context,entity_handle, &nv_public,
+                entity_name);
+            if (!res) {
+                LOG_ERR("Failed reading NV public when calculating name");
+                return false;
+            }
+            break;
+        case TPM2_HT_TRANSIENT:
+        case TPM2_HT_PERSISTENT:
+            res = tpm2_read_transient_persistent_obj_name(sapi_context,
+                    entity_handle, entity_name);
+            if (!res) {
+                LOG_ERR("Failed reading Object public when calculating name");
+                return false;
+            }
+            break;
+        case TPM2_HT_PCR:
+        case TPM2_HT_HMAC_SESSION:
+        case TPM2_HT_POLICY_SESSION:
+        case TPM2_HT_PERMANENT:
+            entity_name->size = sizeof(TPM2_HANDLE);
+            memcpy(temp, (uint8_t *)&entity_handle, sizeof(TPM2_HANDLE));
+            for(unsigned i=0; i<sizeof(TPM2_HANDLE); i++) {
+               entity_name->name[i] = temp[3-i];
+            }
+            break;
+        default:
+            LOG_ERR("Handle not valid for HMAC auth");
+            return false;
+    }
+    return true;
+}
+
+bool command_authentication_hmac(tpm2_session *hmac_session,
+                                 uint8_t *cp_hash,
+                                 TPMS_AUTH_COMMAND *session_data) {
+
+    TPM2B_NONCE nonce_tpm;
+    tpm2_session_get_nonce_tpm(hmac_session, &nonce_tpm);
+    uint8_t *to_hmac_buffer = malloc(SHA256_DIGEST_LENGTH + nonce_tpm.size
+                                + sizeof(TPMA_SESSION));
+    if (!to_hmac_buffer) {
+        LOG_ERR("oom");
+        return false;
+    }
+
+    memcpy(to_hmac_buffer, cp_hash, SHA256_DIGEST_LENGTH);
+
+    memcpy(to_hmac_buffer + SHA256_DIGEST_LENGTH, nonce_tpm.buffer, nonce_tpm.size);
+
+    memcpy(to_hmac_buffer + SHA256_DIGEST_LENGTH + nonce_tpm.size,
+           &session_data->sessionAttributes, sizeof(TPMA_SESSION));
+
+    unsigned int temp_hash_len = 0;
+    HMAC(EVP_sha256(), session_data->hmac.buffer, session_data->hmac.size,
+        to_hmac_buffer, SHA256_DIGEST_LENGTH + nonce_tpm.size + sizeof(TPMA_SESSION),
+        session_data->hmac.buffer, &temp_hash_len);
+    session_data->hmac.size = temp_hash_len;
+    free(to_hmac_buffer);
+
+    return true;
+}
+
+bool tpm2_hmac_auth_get_command_buffer_hmac(TSS2_SYS_CONTEXT *sapi_context,
+    TPM2_HANDLE auth_handle, tpm2_session *hmac_session,
+    TPMS_AUTH_COMMAND *session_data, TPM2B_NAME entity_1_name,
+    TPM2B_NAME entity_2_name) {
+    if (!auth_handle) {
+        LOG_ERR("Authentication handle cannot be NULL");
+        return false;
+    }
+
+    command_authentication_hmac(hmac_session,
+        tpm2_get_cp_hash(sapi_context, entity_1_name, entity_2_name), session_data);
```

Comment:
If `tpm2_get_cp_hash()` fails, `command_authentication_hmac()` will deref NULL. You need to separate these and check for NULL return from `tpm2_get_cp_hash()`

Reply:
Implemented